core/kazoo_number_manager/src/knm_voxbone_util.erl

```diff
@@ -0,0 +1,182 @@
+    <<(api_uri())/binary,"/",Resource/binary,"?",QueryString1/binary>>.
+
+%%------------------------------------------------------------------------------
+%% @doc
+%% @end
+%%------------------------------------------------------------------------------
+-spec default_headers() -> kz_term:proplist().
+default_headers() ->
+    [{"User-Agent", ?KNM_USER_AGENT}
+    ,{"Accept", "application/json"}
+    ,{"Content-Type", "application/json"}
+    ].
+
+%%------------------------------------------------------------------------------
+%% @doc
+%% @end
+%%------------------------------------------------------------------------------
+default_http_options() ->
+    [auth()
+    ,{'ssl', [{'verify', 'verify_none'}]}
+    ,{'timeout', 180 * ?MILLISECONDS_IN_SECOND}
+    ,{'connect_timeout', 180 * ?MILLISECONDS_IN_SECOND}
+    ,{'body_format', 'string'}
+    ].
+
+%%------------------------------------------------------------------------------
+%% @doc resolve the correct api host based on the configured environment
+%% @end
+%%------------------------------------------------------------------------------
+-spec environment_url(kz_term:ne_binary()) -> kz_term:ne_binary().
+environment_url(<<"production">>) -> ?VOXBONE_PRODUCTION_HOST;
+environment_url(<<"sandbox">>) -> ?VOXBONE_SANDBOX_HOST;
+environment_url(<<"beta">>) -> ?VOXBONE_BETA_HOST.
+
+%%------------------------------------------------------------------------------
+%% @doc format to voxbone wildcard pattern
+%% @end
+%%------------------------------------------------------------------------------
+                                                %TODO add  prefix format
+-spec to_voxbone_pattern(knm_phone_number:number()) -> kz_term:ne_binary().
+to_voxbone_pattern(Number) ->
+    <<"%",(knm_phone_number:number(Number))>>.
+
+%%------------------------------------------------------------------------------
+%% @doc sets required querystring params for api requests
+%% @end
+%%------------------------------------------------------------------------------
+-spec required_params() -> qs_options().
+required_params() ->
+    [{'pageNumber', 0}
+    ,{'pageSize', ?VOXBONE_API_PAGE_SIZE}
+    ,{'showEmpty', false}
+    ].
+
+%%------------------------------------------------------------------------------
+%% @doc generic abstraction for voxbone api requests to map to kz_http:req/5
+%% @end
+%%------------------------------------------------------------------------------
+-spec voxbone_request(kz_term:atom(), kz_term:ne_binary(), kz_term:proplist()) -> {'ok', kz_json:object()} | {'error', kz_term:atom()}.
+voxbone_request(Method, Resource, QueryString) ->
+    voxbone_request(Method, Resource, QueryString, []).
+
+-spec voxbone_request(kz_term:atom(), kz_term:ne_binary(), kz_term:proplist(), kz_json:object()) -> {'ok', kz_json:object()} | {'error', kz_term:atom()}.
+voxbone_request(Method, Resource, QueryString, Body) ->
```

Comment:
Body should be `iodata()` or `binary()` no?

Reply:
Or if JObj should be encoded before `kz_http` I think.

Reply:
[kz_http:req/5](https://github.com/2600hz/kazoo/blob/master/core/kazoo_web/src/kz_http.erl#L341) calls kz_term:to_binary in build_request for mutating http verbs which performs the body encode however I will perform a kz_json:encode as to keep things JObj at the voxbone api level.

```erlang
(kazoo_apps@2600Hz-MBP.local)12> T = kz_json:from_list([{'test', 1}, {'test', 2}]). kz_json:encode(T). kz_term:to_binary(T).
{[{test,1},{test,2}]}
(kazoo_apps@2600Hz-MBP.local)13> kz_json:encode(T). kz_term:to_binary(T).
<<"{\"test\":1,\"test\":2}">>
(kazoo_apps@2600Hz-MBP.local)14> kz_term:to_binary(T).
<<"{\"test\":1,\"test\":2}">>
```